### PR TITLE
Use virtualenv in Makefile for python dependencies

### PR DESCRIPTION
**Pull request by cadmic**
`make setup` will create a virtualenv if it doesn't already exist, and all python commands in the Makefile will use the virtualenv without the user having to activate it. This is so the build system can use pip packages without the user having to learn about virtualenv, since that can be hard for beginners.

You will still have to activate the virtualenv if you want to run scripts like `diff.py` by hand, though. I've moved the docs for this out of the main README and into the docs for the tools.

I'm not sure if `distclean` should nuke the virtualenv too. I chose not to here, because I couldn't find a way to automatically deactivate it (`deactivate` it is a shell function, not a command in the virtualenv), and it hopefully it doesn't get too messed up.

(Thanks to kentonm for the idea)

### CODE REVIEW

File: Makefile
Comment (hensldm) [inline Makefile:295]: Thoughts on maybe doing something like this instead
```make
$(VENV): requirements.txt
	python3 -m venv $(VENV)
	$(PYTHON) -m pip install -U pip
	$(PYTHON) -m pip install -U -r requirements.txt
	touch $(VENV)

setup: $(VENV)
	$(MAKE) -C tools
	$(PYTHON) tools/decompress_baserom.py $(VERSION)
	$(PYTHON) extract_baserom.py
	$(PYTHON) extract_assets.py -j$(N_THREADS)
```
this way it keeps all the venv logic together, and would only update if requirements.txt gets updated. This also gives a way to just build/update the venv without needing a full setup again.

**Review by hensldm — COMMENTED** [review]
fwiw, I initially didn't like the idea at first (tbh probably just a personal taste of messing with virtual environments in the Makefile feeling wrong, but not really having any concrete examples why). But thinking/researching it a bit I've started to warm up to the idea.

File: Makefile
Comment (cadmic) [reply Makefile:295]: I thought directories behave badly if you use them as makefile targets, but I do like the idea of splitting out the logic into a separate target, and depending on requirements.txt. Maybe we can have the target be `.venv/pyvenv.cfg` instead, or our own sentinel file?

Comment (cadmic) [reply Makefile:295]: Reading up on this a bit more, I think `setup: | $(VENV)` would work, I'll try it out

Comment (hensldm) [reply Makefile:295]: Personally I'm fine with a sentinel file (would prefer making our own though). Though in that case we may want to add a phony `venv` target that depends on it, just to make setting up/updating the venv without a full setup a little easier (though maybe not necessary).

Comment (cadmic) [reply Makefile]: Okay, I created 2 separate targets, one for creating the venv (which we only want to do once ever) and one for installing python deps if `requirements.txt` changes. I'm also not sure what to call these things so please suggest better names

File: Makefile
Comment (hensldm) [inline Makefile:328]: tbh I don't really see the need to separate this out, since if the venv already exists `python3 -m venv $(VENV)` does nothing so is fine with the others.

Comment (cadmic) [reply Makefile:328]: I think `python3 -m venv $(VENV)` will reinstall the virtualenv, at least I tried editing `.venv/pyvenv.cfg` and ran `python3 -m venv $(VENV)` and it was changed back. It also took about 2 seconds to do so which is pretty slow. I guess the 2 seconds isn't so bad since it will only run when requirements.txt changes, but maybe advanced users want to set up their virtualenv differently (using a different python3 perhaps) and I don't think we should clobber that

Comment (hensldm) [reply Makefile:328]: Hmm, it does seem to clobber `.venv/pyvenv.cfg` but seems to keep pip packages installed.
Perhaps we just go back to using `test -d $(VENV) || python3 -m venv $(VENV)` to see if it already exists. Would kind of like to keep one target as much as possible.

Comment (hensldm) [reply Makefile:328]: Though if we do end up keeping both targets might I suggest `venv` and `venv-deps` as the target names.

File: Makefile
Comment (Dragorn421) [inline Makefile:266]:
```suggestion
all: rom compress
```
(or move this back imo it's just diff/blame noise but whatever)

Comment (cadmic) [reply Makefile:266]: the other PR won so I merged main into this one

File: README.md
Comment (Dragorn421) [inline README.md:117]: I think we should mention the venv step. maybe
```suggestion
This downloads some dependencies (from pip), and compiles tools for the build process.
Then it generates a new ROM "baseroms/gc-eu-mq-dbg/baserom-decompressed.z64" that will have the overdump removed and the header patched.
```

File: Makefile
Comment (cadmic) [reply Makefile]: Oh, I didn't realize I messed it up. ty

**Review by Dragorn421 — APPROVED** [review]

**Review by fig02 — APPROVED** [review]
will wait for anon's approval before merging

**Review by hensldm — APPROVED** [review]

**Pull request merged**
